Add ability to vote for beatmap user tags after completing a play

bdach opened the pull request:
https://github.com/user-attachments/assets/cc18f747-9cc1-47b1-85f1-a535c6c7af6b

---

- [ ] Depends on https://github.com/ppy/osu-web/pull/11999

Closes https://github.com/ppy/osu/issues/31913.

Ended up being a bit of a refactor of `StatisticsPanel` because I hated seeing the inheritance there - most of which was written by yours truly. At least this means there has been personal growth, I guess?

bdach committed:
Rewrite `OverallRanking` to interface directly with `UserStatisticsWatcher`

bdach committed:
Remove inheritance in `StatisticsPanel`

bdach committed:
Add user tag control to results screen's statistics panel

bdach committed:
Show user tag control only following local user gameplay

bdach committed:
Adjust colour of user tag popover control

For some reason in actual gameplay there seems to be an
`OverlayColourProvider` cached that's nowhere to be seen in tests, and
without this change things look bad. Dunno, not looking for it.

bdach committed:
Add visual test coverage of user tag control on results

bdach committed:
Fix xmldoc

peppy committed:
Fix enter to select tag not working in results scren

peppy commented:
Can potentially simplify to this?

```diff
diff --git a/osu.Game.Tests/Visual/Ranking/TestSceneResultsScreen.cs b/osu.Game.Tests/Visual/Ranking/TestSceneResultsScreen.cs
index b19288fd99..4758b70526 100644
--- a/osu.Game.Tests/Visual/Ranking/TestSceneResultsScreen.cs
+++ b/osu.Game.Tests/Visual/Ranking/TestSceneResultsScreen.cs
@@ -404,7 +404,7 @@ public TestResultsScreen(ScoreInfo score)
                 : base(score)
             {
                 AllowRetry = true;
-                ShowUserStatistics = true;
+                IsLocalPlay = true;
             }
 
             protected override void LoadComplete()
diff --git a/osu.Game.Tests/Visual/Ranking/TestSceneStatisticsPanel.cs b/osu.Game.Tests/Visual/Ranking/TestSceneStatisticsPanel.cs
index a64fd488bc..02d30d12e6 100644
--- a/osu.Game.Tests/Visual/Ranking/TestSceneStatisticsPanel.cs
+++ b/osu.Game.Tests/Visual/Ranking/TestSceneStatisticsPanel.cs
@@ -214,7 +214,6 @@ public void TestTagging()
                         State = { Value = Visibility.Visible },
                         Score = { Value = score },
                         AchievedScore = score,
-                        ShowUserTagControl = true,
                     }
                 };
             });
diff --git a/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs b/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs
index 67a67cf271..3d4b46f49e 100644
--- a/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs
+++ b/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs
@@ -198,13 +198,11 @@ protected override ResultsScreen CreateResults(ScoreInfo score)
             return multiplayerLeaderboard.TeamScores.Count == 2
                 ? new MultiplayerTeamResultsScreen(score, Room.RoomID.Value, PlaylistItem, multiplayerLeaderboard.TeamScores)
                 {
-                    ShowUserStatistics = true,
-                    ShowUserTagControl = true,
+                    IsLocalPlay = true,
                 }
                 : new MultiplayerResultsScreen(score, Room.RoomID.Value, PlaylistItem)
                 {
-                    ShowUserStatistics = true,
-                    ShowUserTagControl = true,
+                    IsLocalPlay = true,
                 };
         }
 
diff --git a/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsPlayer.cs b/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsPlayer.cs
index 80b378bdcf..dc4078cb1f 100644
--- a/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsPlayer.cs
+++ b/osu.Game/Screens/OnlinePlay/Playlists/PlaylistsPlayer.cs
@@ -59,8 +59,7 @@ protected override ResultsScreen CreateResults(ScoreInfo score)
             return new PlaylistItemScoreResultsScreen(score, Room.RoomID.Value, PlaylistItem)
             {
                 AllowRetry = true,
-                ShowUserStatistics = true,
-                ShowUserTagControl = true,
+                IsLocalPlay = true,
             };
         }
 
diff --git a/osu.Game/Screens/Play/SubmittingPlayer.cs b/osu.Game/Screens/Play/SubmittingPlayer.cs
index 04cf473173..dc3e5f08ac 100644
--- a/osu.Game/Screens/Play/SubmittingPlayer.cs
+++ b/osu.Game/Screens/Play/SubmittingPlayer.cs
@@ -328,8 +328,7 @@ private Task submitScore(Score score)
         protected override ResultsScreen CreateResults(ScoreInfo score) => new SoloResultsScreen(score)
         {
             AllowRetry = true,
-            ShowUserStatistics = true,
-            ShowUserTagControl = true,
+            IsLocalPlay = true,
         };
     }
 }
diff --git a/osu.Game/Screens/Ranking/ResultsScreen.cs b/osu.Game/Screens/Ranking/ResultsScreen.cs
index fcf90a3e28..6da731588f 100644
--- a/osu.Game/Screens/Ranking/ResultsScreen.cs
+++ b/osu.Game/Screens/Ranking/ResultsScreen.cs
@@ -79,13 +79,10 @@ public abstract partial class ResultsScreen : ScreenWithBeatmapBackground, IKeyB
         public bool AllowWatchingReplay { get; init; } = true;
 
         /// <summary>
-        /// Whether the user's personal statistics should be shown on the extended statistics panel
-        /// after clicking the score panel associated with the <see cref="Score"/> being presented.
-        /// Requires <see cref="Score"/> to be present.
+        /// Whether the provided score is for a local user's play.
+        /// This will trigger elements like the user's ranking to display.
         /// </summary>
-        public bool ShowUserStatistics { get; init; }
-
-        public bool ShowUserTagControl { get; init; }
+        public bool IsLocalPlay { get; init; }
 
         private Sample? popInSample;
 
@@ -127,8 +124,7 @@ private void load(AudioManager audio)
                                         {
                                             RelativeSizeAxes = Axes.Both,
                                             Score = { BindTarget = SelectedScore },
-                                            AchievedScore = ShowUserStatistics && Score != null ? Score : null,
-                                            ShowUserTagControl = ShowUserTagControl,
+                                            AchievedScore = IsLocalPlay && Score != null ? Score : null,
                                         },
                                         ScorePanelList = new ScorePanelList
                                         {
diff --git a/osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs b/osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs
index 16de00fcf1..0cee01cf77 100644
--- a/osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs
+++ b/osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs
@@ -37,11 +37,6 @@ public partial class StatisticsPanel : VisibilityContainer
         /// </summary>
         public ScoreInfo? AchievedScore { get; init; }
 
-        /// <summary>
-        /// Whether to show a control that allows to assign user tags to the played beatmap.
-        /// </summary>
-        public bool ShowUserTagControl { get; init; }
-
         protected override bool StartHidden => true;
 
         [Resolved]
@@ -235,7 +230,7 @@ protected virtual IEnumerable<StatisticItem> CreateStatisticItems(ScoreInfo newS
                 });
             }
 
-            if (ShowUserTagControl
+            if (AchievedScore != null
                 && newScore.BeatmapInfo!.OnlineID > 0
                 && api.IsLoggedIn)
             {

```

peppy commented:
I'd also probably want to see this applied (currently based on top of above diff):

```diff
diff --git a/osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs b/osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs
index 0cee01cf77..2481b0560f 100644
--- a/osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs
+++ b/osu.Game/Screens/Ranking/Statistics/StatisticsPanel.cs
@@ -14,6 +14,7 @@
 using osu.Framework.Graphics.Containers;
 using osu.Framework.Input.Events;
 using osu.Game.Beatmaps;
+using osu.Game.Graphics;
 using osu.Game.Graphics.Containers;
 using osu.Game.Graphics.UserInterface;
 using osu.Game.Online.API;
@@ -234,12 +235,29 @@ protected virtual IEnumerable<StatisticItem> CreateStatisticItems(ScoreInfo newS
                 && newScore.BeatmapInfo!.OnlineID > 0
                 && api.IsLoggedIn)
             {
-                yield return new StatisticItem("Tag the beatmap!", () => new UserTagControl(newScore.BeatmapInfo)
+                if (
+                    // We may want to iterate on this condition
+                    AchievedScore.Rank >= ScoreRank.C
+                )
                 {
-                    RelativeSizeAxes = Axes.X,
-                    Anchor = Anchor.Centre,
-                    Origin = Anchor.Centre,
-                });
+                    yield return new StatisticItem("Tag the beatmap!", () => new UserTagControl(newScore.BeatmapInfo)
+                    {
+                        RelativeSizeAxes = Axes.X,
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                    });
+                }
+                else
+                {
+                    yield return new StatisticItem("Tag the beatmap!", () => new OsuTextFlowContainer(cp => cp.Font = OsuFont.GetFont(size: StatisticItem.FONT_SIZE, weight: FontWeight.SemiBold))
+                    {
+                        RelativeSizeAxes = Axes.X,
+                        AutoSizeAxes = Axes.Y,
+                        Anchor = Anchor.Centre,
+                        Origin = Anchor.Centre,
+                        Text = " Set a better score to contribute to beatmap tags!",
+                    });
+                }
             }
         }
 

```

peppy reviewed (commented):
As commented.

bdach committed:
Simplify boolean flags

bdach committed:
Require high enough score rank for tagging beatmap

bdach commented:
> Can potentially simplify to this?

Applied verbatim in https://github.com/ppy/osu/pull/32346/commits/f78e371b1b5194cd7ee74e79ef54926d6ac7f3a4

> I'd also probably want to see this applied (currently based on top of above diff):

Applied with alterations in https://github.com/ppy/osu/pull/32346/commits/5b1a4c8ed14bbf17baa11c87325226540a5fe170; I added a test case, and you also specified `Anchor` & `Origin` as `Centre` on the text flow which doesn't work if you wanted to centre the text (you want `TextAnchor` for that, and that's what I changed there).

bdach committed:
Fix test

bdach committed:
Switch add beatmap tag request to using path paramx

See https://github.com/ppy/osu-web/pull/11999.

mixalisnew commented:
if a map was created for a tourny or the creator wanted to add it later could we add HR1 for example if said beatmap (notmapset) was h1 in the pool?

peppy commented:
> if a map was created for a tourny or the creator wanted to add it later could we add HR1 for example if said beatmap (notmapset) was h1 in the pool?

I think you're in the wrong place, or misunderstanding what this is.
